sotodlib/hwp/hwp.py

```diff
@@ -6,429 +6,524 @@
+    modes : list of int
+        The frequencies of the sine and cosine basis functions to use.
 
-    Args
-    -----
-        aman: AxisManager
-            AxisManager including target timestream data
-        signal: str
-            name of target timestream data in AxisManager
-        name: str
-            output timestram name into AxisManager
-        mode: int
-            0 = 2f&4f, 1 = 2f only, 2 = 4f only, 3 = all up to 5f
-        prefit_cutoff: float
-            cut off value of pre-lowpass filter
-        bins: int
-            number of bins for hwpss histogram
+    Returns
+    -------
+    fitsig : numpy.ndarray
+        The fitted signal, obtained by evaluating the model with the optimal coefficients.
+    coeffs : numpy.ndarray
+        The coefficients of the sine and cosine basis functions that best fit the data.
+    covars : numpy.ndarray
+        The covariance matrix of the coefficients, estimated from the data errors.
+    redchi2s : numpy.ndarray
+        The reduced chi-square statistic of the fit, computed for each data point.
     """
+    vects = np.zeros([2*len(modes), x.shape[0]], dtype='float32')
+    for i, mode in enumerate(modes):
+        vects[2*i, :] = np.sin(mode*x)
+        vects[2*i+1, :] = np.cos(mode*x)
 
-    filt = tod_ops.filters.high_pass_sine2(cutoff=prefilt_cutoff)
-    signal_prefilt = np.array(tod_ops.fourier_filter(
-        aman, filt, detrend=None, signal_name=signal))
+    I = np.linalg.inv(np.tensordot(vects, vects, (1, 1)))
+    coeffs = np.matmul(ys, vects.T)
+    coeffs = np.dot(I, coeffs.T).T
+    fitsig = np.matmul(vects.T, coeffs.T).T
 
-    hwpss_denom = np.histogram(aman.hwp_angle, bins=bins, range=[0, 2 * np.pi])[0]
-    hwp_angle_bins = np.linspace(0, 2 * np.pi, bins)
-    hwpss = []
-    for i in range(aman.dets.count):
-        hwpss.append(np.histogram(aman.hwp_angle,bins=bins,range=[0,2*np.pi],
-                     weights=signal_prefilt[i])[0]/np.where(hwpss_denom==0, 1, hwpss_denom))
-        idx = np.argwhere(hwpss[i] != 0).flatten()
-        if i == 0:
-            hwp_angle_bins = hwp_angle_bins[idx]
-        hwpss[i] = hwpss[i][idx]
-
-    if name in aman.keys():
-        aman.move(name, None)
-    aman.wrap_new(name, ('dets', 'samps'))
-    if mode == 0:
-        func = func_hwpss_2f4f
-    elif mode == 1:
-        func = func_hwpss_2f
-    elif mode == 2:
-        func = func_hwpss_4f
-    elif mode == 3:
-        func = func_hwpss_all
-    for i in range(aman.dets.count):
-        popt, pcov = curve_fit(func, hwp_angle_bins, hwpss[i], maxfev=1000000)
-        aman[name][i] = func(aman.hwp_angle, *popt)
+    # covariance of coefficients
+    covars = np.zeros((ys.shape[0], 2*len(modes), 2*len(modes)))
+    for det_idx in range(ys.shape[0]):
+        covars[det_idx, :, :] = I * yerrs[det_idx]**2
 
+    # reduced chi-square
+    redchi2s = np.sum(
+        ((ys - fitsig)/yerrs[:, np.newaxis])**2, axis=-1) / (x.shape[0] - 2*len(modes))
```

Comment:
I found the same issue you show here. I addressed this by setting all bins with counts = 0 to nan and then masking nan values in the computations later (fitting, linear regression, residual estimation, etc.). I don't get the error now and the fitted values are reasonable even for just a simple sine wave input.

Reply:
Ah, that is a great catch. But let me comments about that.

First comment is you see this issue because you are using simulation data. The simulation has exactly same dt_sampling and exactly steady HWP rotation. So when you bin the signal, bins **only** at a greatest common divisor of sampling rate and rotation speed. That is, the HWPSS is **extremely** coherent for current simulation data.
In real data, rotation speed varies and this issue rarely happens with real data.
That is why I did not found this bug, as I validated with real data.

But I agree with that we should check number of bins are proper or not.
So I will add a function to derive a proper number of bins. And will raise error if binning is too dense.

And I am sorry, but I would like to undo Max's commit, setting bins with counts = 0 to nan.
The reduced chi2s are wrongly calculated:
redchi2 = ( ( y_fit(bin) - y_data(bin)) / sigma )^2 / dof
, where y_fit has full samples and y_data(bin) dropping at some bins, and returning extremely high redchi2s.

Does it make sense?

Reply:
Thank you, Max. Yeah, that is exactly what I thought of. And I have committed that.

Reply:
This has been addressed with raising a warning in logs, returning nans for 0 count bins, and adjusting all calculation functions to mask bins w/ zero counts so I will resolve this comment.